### ISSUE TITLE
Azure Storage: ContentHash not calculated after UploadAsync

### ISSUE THREAD

**Issue by filipCristian** [Azure/azure-sdk-for-net#14037]
Hello all,

I will start by saying: Not sure if its a bug but at this point I'm afraid to ask. To moderators, please feel free to edit the issue type accordingly.

My project was running v9 of WindowsAzure.Storage which is deprecated and I decided to upgrade it to this sdk, but I have run into the following problem.
After uploading a blob, the content hash was not calculated, nor the _BlobContentInfo_ or _BlobProperties_ were returning content hash and my whole flow is base on receiving the hash from Azure. After some good amount of time spend debugging, I found out that the stream was the problem.

I am taking the stream from _HttpRequest_ using the method _GetBufferlessInputStream_ and if I upload this stream the content hash is not calculated, even if I go into azure storage explorer, the ContentMD5 of the blob is empty. BUT, but if I take the stream from _HttpRequest_ using _InputStream_ everything works as expected.

Do you know why this different behavior? And do you know how to make to receive content hash for streams received by _GetBufferlessInputStream_ method.

P.S. I think its obvious, but with the old sdk, content hash was calculated for streams received by _GetBufferlessInputStream_ method

**Comment by jsquire** [comment]
Thank you for your feedback.  Tagging and routing to the team best able to assist.

**Comment by filipCristian** [comment]
So the code looks like this:

`var stream = HttpContext.Current.Request.GetBufferlessInputStream(disableMaxRequestLength: true)

var container = _blobServiceClient.GetBlobContainerClient(containerName);

// get blob reference
var blob = container.GetBlockBlobClient(blobPath);

BlobHttpHeaders blobHttpHeaders = null;
if (!string.IsNullOrWhiteSpace(fileContentType))
{
     blobHttpHeaders = new BlobHttpHeaders()
     {
          ContentType = fileContentType,
     };
}

// retry already configured of Azure Storage API
await blob.UploadAsync(stream, httpHeaders: blobHttpHeaders);

return await blob.GetPropertiesAsync();`

In the code snippet from above `ContentHash `is NOT calculated, but if I change the way I am getting the stream from the http request with following snippet `ContentHash `is calculated.

`var stream = HttpContext.Current.Request.InputStream`

**Comment by amnguye** [comment]
When providing `ContentHash` alongside `Upload`, This hash is used to verify the integrity of the blob during transport.  When this header (`BlobHttpHeaders.ContentHash`) is specified, the storage service checks the hash that has arrived with the one that was sent. The service sends a  400 (Bad Request) if they do not match.

That's strange behavior that depending on the stream the ContentHash value would get returned or not when it should be returned only when you set the ContentHash to check against the service's hash value. 

Just for clarification you don't get back a 400?

**Comment by filipCristian** [comment]
No, I do NOT get 400. The response code returned by `UploadAsync` is always 201.
And in both cases, I do NOT specify the header `BlobHttpHeaders.ContentHash`.

**Comment by TiTi** [comment]
I'm also having this issue when copying blobs between 2 differents storage accounts like so:
```
BlobClient sourceBlob = containerSource.GetBlobClient(blobName);
//sourceBlob.Exists() == true
BlobClient destBlob = containerDestination.GetBlobClient(blobName);
//destBlob.Exists() == false

BlobProperties props = sourceBlob.GetProperties().Value;
BlobDownloadInfo infos = sourceBlob.Download();
containerDestination.UploadBlob(blobName, infos.Content);
destBlob.SetMetadata(props.Metadata);
destBlob.SetAccessTier(AccessTier.Archive);
```

**Comment by nickalbrecht** [comment]
Ran into this today. From my digging, it appears this is a symptom of the type of `Stream` you use to upload, and it's not really a bug. In order to generate a hash for your blob (which is done on the client side before uploading by the looks of it), it needs to read the stream. Which means it would need to reset the position of your stream back to 0 (for the actual upload process) after generating the hash. Doing this requires the ability to perform the Seek operation on the stream. If your stream doesn't support Seek, then it looks like it doesn't generate the hash.

To get around the issue, make sure the stream you provide supports Seek (`CanSeek`). If it doesn't, then use a different Stream/copy your data to a stream that does (for example `MemoryStream`). The alternative would be for the internals of the Blob SDK to do this for you.

**Comment by seanmcc-msft** [comment]
I'm going to close this issue, please re-open if there are further questions.

**Comment by nickalbrecht** [comment]
At the very least, I'd like to see this caveat added to the documentation. It's worth mentioning (especially for those upgrading from the previous version of the SDK since it didn't have this limitation) that if the `Stream` provided does not support Seek, that the upload still work - but that the `ContentHash` logic will be skipped and left blank in order to conserve the single read operation for the actual actual upload.

**Comment by NeilMacMullen** [comment]
@seanmcc-msft   I don't think this issue should have been closed.  :-) 



The original report implied that the content-hash was not being stored on the server copy.  The reply from @amnguye appears to be talking about a different mechanism whereby the client can (optionally) supply a hash to check the integrity of the transfer.    I get that this might be part of the http layer API but why would it ever be exposed in the C# library - users just want to do an UploadAsync without any further ceremony and to be able to rely on it just working.  

So there are a couple of issues here...

1) If the server is capable of generating a hash why would it not always store it with the blob regardless of whether the client has requested an integrity check  (arguably that's a bug/question for the Azure infrastructure team).

2) Why would the client library ever NOT send the hash?  I'm sure there must be an implementation that allows the hash to be calculated as the data is streamed to avoid requiring a seekable Stream.

I have raised a related issue at https://github.com/Azure/azure-sdk-for-net/issues/17676 - in this case the hash appears not to be generated for large blobs (or maybe when the blob is uploaded in concurrent blocks).  In any case it seems reasonable to expect a blob to have a few guaranteed properties - name, length, hash and contents and for those properties to be present (and consistent) no matter what client was used to perform the upload.
